internal/metrics/metrics.go

```diff
@@ -1,17 +1,224 @@
 package metrics
 
 import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"log"
 	"net/http"
+	"time"
 
+	"github.com/canonical/gocert/internal/certdb"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/collectors"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
-// Returns an HTTP handler for Prometheus metrics.
-func NewPrometheusMetricsHandler() http.Handler {
-	reg := prometheus.NewRegistry()
-	reg.MustRegister(collectors.NewGoCollector(), collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
-	prometheusHandler := promhttp.HandlerFor(reg, promhttp.HandlerOpts{})
-	return prometheusHandler
+const ONE_SECOND = 1000000000
+
+type PrometheusMetrics struct {
+	http.Handler
+	registry                       *prometheus.Registry
+	CertificateRequests            prometheus.Gauge
+	OutstandingCertificateRequests prometheus.Gauge
+	Certificates                   prometheus.Gauge
+	CertificatesExpiringIn1Day     prometheus.Gauge
+	CertificatesExpiringIn7Days    prometheus.Gauge
+	CertificatesExpiringIn30Days   prometheus.Gauge
+	CertificatesExpiringIn90Days   prometheus.Gauge
+	ExpiredCertificates            prometheus.Gauge
+
+	RequestsTotal    prometheus.CounterVec
+	RequestsDuration prometheus.HistogramVec
+}
+
+// NewMetricsSubsystem returns the metrics endpoint HTTP handler and the Prometheus metrics for the server and middleware.
+func NewMetricsSubsystem(db *certdb.CertificateRequestsRepository) *PrometheusMetrics {
+	csrs, err := db.RetrieveAll()
+	if err != nil {
+		log.Println(errors.Join(errors.New("error generating metrics repository: "), err))
+	}
+	metricsBackend := newPrometheusMetrics()
+	metricsBackend.generateMetrics(csrs)
+	go func() {
+		time.Sleep(120 * ONE_SECOND)
+		metricsBackend.generateMetrics(csrs)
+	}()
+	metricsBackend.Handler = promhttp.HandlerFor(metricsBackend.registry, promhttp.HandlerOpts{})
+	return metricsBackend
+}
```

Comment:
Should we use a ticker instead of continuously running the goroutine?

Ex:

```go
ticker := time.NewTicker(120 * time.Second)
go func() {
    for range ticker.C {
        metricsBackend.generateMetrics(csrs)
    }
}()
```

This will also remove the need of defining and using the `ONE_SECOND` const.

Reply:
Tickers look nice, I've changed it to use this.